### PR TITLE
Migrate update_optimize.test.lua to luatest

### LINKED ISSUE #6556: Migrate vinyl/update_optimize.test.lua
Migrate  vinyl/update_optimize.test.lua to luatest.

As a part of tarantool/test-run#304

### PR DESCRIPTION
**Pull request by VitaliyaIoffe**
update_optimize.test.lua was migrated to the luatest framework. Also,
deleted old version.

Close #6556

### CODE REVIEW

File: test/vinyl-luatest/update_optimize_test.lua [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+
+local function check_box_snapshot(res)
+    t.assert_equals(g.server:eval("return box.snapshot()"), res)
+end
+
+local function insert(...)
+    return space:insert(...)
+end
+local function check_insert(data, expected)
+    t.assert_equals(g.server:exec(insert, data), expected)
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply test/vinyl-luatest/update_optimize_test.lua:54]: I think we will fail.

File: test/vinyl-luatest/update_optimize_test.lua [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+    g.server:eval("space = box.schema.space.create('test', { engine = 'vinyl' })")
+    g.server:eval([[
+         dumped_stmt_count = function(indices)
+            local dumped_count = 0
+            for _, index in ipairs(indices) do
+                dumped_count = dumped_count + index:stat().disk.dump.output.rows
+            end
+            return dumped_count
+         end
+    ]])
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply test/vinyl-luatest/update_optimize_test.lua]: Some tests are rewritten via `exec`. Please take a look at the approach.

File: test/vinyl-luatest/update_optimize_test.lua [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+        vinyl_run_size_ratio = 2,
+        vinyl_cache = 10240, -- 10kB
+        vinyl_max_tuple_size = 1024 * 1024 * 6,
+    }
+    g.server = Server:new({alias = 'master', box_cfg = boxcfg} )
+    g.server:start()
+end
+
+g.before_each(function()
+    g.server:eval("space = box.schema.space.create('test', { engine = 'vinyl' })")
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply test/vinyl-luatest/update_optimize_test.lua]: But in this case, we need to drop space in the test as well which is not very convenient.

File: test/vinyl-luatest/update_optimize_test.lua [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+		fiber.sleep(0)
+	end
+	return index:stat().disk.dump.count
+end;
+
+g.test_optimize_one_index = function()
+    -- optimize one index
+    g.server:eval("index = space:create_index('primary', { run_count_per_level = 20 })")
+    g.server:eval([[index2 = space:create_index('secondary',
+        { parts = {5, 'unsigned'}, run_count_per_level = 20 })]])
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply test/vinyl-luatest/update_optimize_test.lua]: Maybe in the future :)

File: .luacheckrc [hunk 5]
```diff
@@ -111,3 +111,10 @@ files["test/replication/replica_quorum.lua"] = {
         "nonexistent_uri",
     },
 }
+
+files["test/vinyl-luatest/update_optimize_test.lua"] = {
+    globals = {
+        "space",
+        "index",
+    },
+}
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply .luacheckrc]: For now, it's not needed anymore. Removed.

File: test/vinyl-luatest/update_optimize_test.lua [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+    check_update({{2}, {{'=', 7, 100}, {'+', 5, 10},
+        {'#', 3, 1}}}, {11, 2, 13, 24, 15, 100}
+    ) -- change two cols but then move range with all indexed fields
+    check_box_snapshot('ok')
+    local dump_count = g.server:exec(wait_for_dump, {dump_count})
+    new_stmt_count = g.server:eval("return dumped_stmt_count({index, index2, index3})")
+    t.assert_equals(new_stmt_count - old_stmt_count, 15)
+    old_stmt_count = new_stmt_count
+    t.assert_equals(
+        g.server:eval("return space:select{}"), {{11, 2, 13, 24, 15, 100},
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply test/vinyl-luatest/update_optimize_test.lua]: Fixed.

File: test/vinyl-luatest/update_optimize_test.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (locker) [reply test/vinyl-luatest/update_optimize_test.lua]: After I [found out](https://github.com/tarantool/luatest/issues/189#issuecomment-986822525) that it's possible to write `t.assert` in `exec`, I don't think we need this at all.

File: test/vinyl-luatest/update_optimize_test.lua [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+    local addr = g.server:eval("return box.cfg.listen")
+    local conn = net_box.connect(addr)
+    local stream = conn:new_stream()
+    stream:begin()
+    local space = stream.space.test
+    local index = stream.space.test.index.primary
+    t.assert_equals(space:replace{20, 200, 2000, 20000, 200000, 2000000},
+        {20, 200, 2000, 20000, 200000, 2000000})
+    t.assert_equals(index:update({200}, {{'=', 6, 2}}), {20, 200, 2000, 20000, 200000, 2})
+    stream:commit()
```
[the remark this answers is not in the document]
Comment (locker) [reply test/vinyl-luatest/update_optimize_test.lua]: Now I think we should [wrap](https://github.com/tarantool/luatest/issues/189#issuecomment-986822525) the whole test in an `exec`. Not relevant anymore.

**Review by locker — CHANGES_REQUESTED** [review]
Please squash all commits in this PR or create a new PR: I always reviews PRs commit-by-commit, and the option to squash and merge is disabled in this repository.

File: test/vinyl-luatest/update_optimize_test.lua [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+        vinyl_range_size = 1024 * 64,
+        vinyl_page_size = 1024,
+        vinyl_run_count_per_level = 1,
+        vinyl_run_size_ratio = 2,
+        vinyl_cache = 10240, -- 10kB
+        vinyl_max_tuple_size = 1024 * 1024 * 6,
+    }
+    g.server = Server:new({alias = 'master', box_cfg = boxcfg} )
+    g.server:start()
+end
```
[the remark this answers is not in the document]
Comment (ylobankov) [reply test/vinyl-luatest/update_optimize_test.lua]: Please describe what you mean. It is not clear to me.

[the remark this answers is not in the document]
Comment (locker) [reply test/vinyl-luatest/update_optimize_test.lua]: The value of `box_cfg` was copied from [test/vinyl/vinyl.lua](https://github.com/tarantool/tarantool/blob/master/test/vinyl/vinyl.lua) which was used by all old tests. We could put this default config in a new Lua file inside test/vinyl-luatest and use it in all new tests so as not to copy-paste it every time we add/migrated a Vinyl test.

[the remark this answers is not in the document]
Comment (ylobankov) [reply test/vinyl-luatest/update_optimize_test.lua]: Fixed.

File: test/vinyl-luatest/box_cfg.lua [hunk 9]
```diff
@@ -0,0 +1,14 @@
+return {
+    work_dir = os.environ()['VARDIR'],
```
Comment (locker) [inline test/vinyl-luatest/box_cfg.lua:2]: The working directory should be set by luatest (isn't it already?).

File: test/vinyl-luatest/box_cfg.lua [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    memtx_memory = 512 * 1024 * 1024,
+    memtx_max_tuple_size = 4 * 1024 * 1024,
```
Comment (locker) [inline test/vinyl-luatest/box_cfg.lua:4]: Let's drop memtx parameters from the config - the test should work just fine with the defaults. After all, we test vinyl here, not memtx.

File: test/vinyl-luatest/suite.ini [hunk 11]
```diff
@@ -1,4 +1,4 @@
 [default]
 core = luatest
-description = vinyl space engine luatests
+description = vinyl integration tests in luatest
```
Comment (locker) [inline test/vinyl-luatest/suite.ini:3]: Please avoid unrelated changes - they complicate review and obscure the history.

File: test/vinyl-luatest/update_optimize_test.lua [hunk 12]
```diff
@@ -0,0 +1,530 @@
...
+    local dumped_count = 0
+    for _, i in ipairs(indexes) do
+        dumped_count = dumped_count +
+            box.space.test.index[i]:stat().disk.dump.output.rows
+    end
+    return dumped_count
+end]]
+
+g.before_all(function()
+    local box_cfg = require('test.vinyl-luatest.box_cfg')
```
Comment (locker) [inline test/vinyl-luatest/update_optimize_test.lua:17]: This looks weird Lua-wise IMO. Let's instead name this file `common.lua` and put the default config in `common.default_box_cfg`.

Also, I'd make `default_box_cfg` a function so that a test could update the config in place, for example:

```lua
local common = require('test.vinyl.common')
local box_cfg = common.default_box_cfg()
box_cfg.vinyl_cache = 1024 * 1024
```

File: test/vinyl-luatest/update_optimize_test.lua [hunk 13]
```diff
@@ -0,0 +1,530 @@
...
+    g.server = server:new({alias = 'master', box_cfg = box_cfg})
+    g.server:start()
+    g.server:eval(dump_stmt_count_str)
```
Comment (locker) [inline test/vinyl-luatest/update_optimize_test.lua:20]: Please use `exec` here:

```lua
g.server:exec(function()
    rawset(_G, 'dump_stmt_count', function(indexes)
        ...
    end)
end)
```

File: test/vinyl-luatest/update_optimize_test.lua [hunk 14]
```diff
@@ -0,0 +1,530 @@
...
+    g.server:exec(function()
+        local t = require('luatest')
+
+        box.space.test:create_index('primary', {run_count_per_level = 20})
+        box.space.test:create_index('secondary',
+            {parts = {5, 'unsigned'}, run_count_per_level = 20})
+
+        box.snapshot()
+
+        local old_stmt_count = dump_stmt_count({'primary', 'secondary'})
```
Comment (locker) [inline test/vinyl-luatest/update_optimize_test.lua:43]: Please access globals via `_G` and drop the luacheck exception:

```lua
local old_stmt_count = _G.dump_stmt_count({'primary', 'secondary'})
```

**Review by locker — CHANGES_REQUESTED** [review]

File: test/vinyl-luatest/update_optimize_test.lua (hunk 14 above)
Comment (ylobankov) [reply test/vinyl-luatest/update_optimize_test.lua]: Fixed.

File: test/vinyl-luatest/update_optimize_test.lua (hunk 13 above)
Comment (ylobankov) [reply test/vinyl-luatest/update_optimize_test.lua]: Thanks! Fixed.

File: test/vinyl-luatest/suite.ini (hunk 11 above)
Comment (ylobankov) [reply test/vinyl-luatest/suite.ini]: Fixed.

**Review by locker — APPROVED** [review]
